src/Razor/src/Microsoft.CodeAnalysis.Razor.CohostingShared/Hover/CohostHoverEndpoint.cs

```diff
@@ -54,29 +55,87 @@ public ImmutableArray<Registration> GetRegistrations(VSInternalClientCapabilitie
                 cancellationToken)
             .ConfigureAwait(false);
 
-        if (response.Result is LspHover hover)
+        if (razorResponse.StopHandling)
         {
-            return hover;
+            return razorResponse.Result;
         }
 
-        if (response.StopHandling)
-        {
-            return null;
-        }
-
-        return await _requestInvoker.MakeHtmlLspRequestAsync<TextDocumentPositionParams, LspHover>(
+        var htmlHover = await _requestInvoker.MakeHtmlLspRequestAsync<TextDocumentPositionParams, LspHover>(
             razorDocument,
             Methods.TextDocumentHoverName,
             request,
             cancellationToken).ConfigureAwait(false);
+
+        return MergeHtmlAndRazorHoverResponses(razorResponse.Result, htmlHover);
+    }
+
+    private static LspHover? MergeHtmlAndRazorHoverResponses(LspHover? razorHover, LspHover? htmlHover)
+    {
+        if (razorHover is null)
+        {
+            return htmlHover;
+        }
+
+        if (htmlHover is null
+            || htmlHover.Range != razorHover.Range)
+        {
+            return razorHover;
+        }
+
+        var htmlStringResponse = htmlHover.Contents.Match(
+            static s => s,
+            static markedString => null,
+            static stringOrMarkedStringArray => null,
+            static markupContent => markupContent.Value
+        );
+
+        if (htmlStringResponse is not null)
+        {
+            // This logic is to prepend HTML hover content to the razor hover content if both exist.
+            // The razor content comes through as a ContainerElement, while the html content comes
+            // through as MarkupContent. We need to extract the html content and insert it at the
+            // start of the combined ContainerElement.
+            if (razorHover is VSInternalHover razorVsInternalHover
+                && razorVsInternalHover.RawContent is ContainerElement razorContainerElement)
+            {
+                var htmlStringClassifiedTextElement = ClassifiedTextElement.CreatePlainText(htmlStringResponse);
+                var verticalSpacingTextElement = ClassifiedTextElement.CreatePlainText(string.Empty);
+                var htmlContainerElement = new ContainerElement(
+                    ContainerElementStyle.Stacked,
+                    [htmlStringClassifiedTextElement, verticalSpacingTextElement]);
+
+                // Modify the existing hover's RawContent to prepend the HTML content.
+                razorVsInternalHover.RawContent = new ContainerElement(razorContainerElement.Style, [htmlContainerElement, .. razorContainerElement.Elements]);
+            }
+            else
+            {
+                var razorStringResponse = razorHover.Contents.Match(
+                    static s => s,
+                    static markedString => null,
+                    static stringOrMarkedStringArray => null,
```

Comment:
Should probably throw for these two, since this is only matching content that we produce, so if we ever start producing it some tests will fail and we'll know to add support here.